### PR TITLE
Added testcase:NamespaceLabelledBackupShareWithDifferentAccessMode:PA-749

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:
Added new test case for 'Share the namespace labeled backups with different access modes'
**Which issue(s) this PR fixes** (optional)
Closes #PA-749: https://portworx.testrail.net/index.php?/tests/view/22623516&group_by=cases:title&group_order=asc&group_id=-1

**Special notes for your reviewer**:
CreateBackupWithNamespaceLabel, mapToString,GenerateRandomLabels,NamespaceLabelBackupSuccessCheck will be removed from my PR once https://github.com/portworx/torpedo/pull/1248 is merged.

BYOC link: https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/810/consoleFull

https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/943/consoleFull

Latest: https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/1069/consoleFull

### CODE REVIEW

File: tests/backup/namespace_labelled_backup_test.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+		log.Infof("Cleaning up users")
+		for _, userName := range users {
+			wg.Add(1)
+			go func(userName string) {
+				defer wg.Done()
+				err := backup.DeleteUser(userName)
+				dash.VerifySafely(err, nil, fmt.Sprintf("Deleting user %v", userName))
+			}(userName)
+		}
+		wg.Wait()
```
Comment (skonda-px) [inline tests/backup/namespace_labelled_backup_test.go:194]: for this also create a ticket for cleanup of users/restores.
and deleting restore namespaces.

**Review by skonda-px — CHANGES_REQUESTED** [review]

File: tests/backup/namespace_labelled_backup_test.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
+package tests
```
Comment (skonda-px) [inline tests/backup/namespace_labelled_backup_test.go:1]: Change the file name to backup_namespace_labelled_test.go

File: tests/backup/namespace_labelled_backup_test.go [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+	)
+	userContextsList := make([]context.Context, 0)
+	accessUserBackupContext := make(map[userAccessContext]string)
+	userRestoreContext := make(map[context.Context]string)
+	numberOfUsers := 3
+	backupLocationMap := make(map[string]string)
+	users := make([]string, 0)
+	labelSelectors := make(map[string]string)
+	bkpNamespaces := make([]string, 0)
+	labels := make(map[string]string)
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/namespace_labelled_backup_test.go:40]: done

File: tests/backup/namespace_labelled_backup_test.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+				cloudCredName = fmt.Sprintf("%s-%s-%v", "cred", provider, time.Now().Unix())
+				bkpLocationName = fmt.Sprintf("autogenerated-backup-location-%v", time.Now().Unix())
+				cloudCredUID = uuid.New()
+				backupLocationUID = uuid.New()
+				backupLocationMap[backupLocationUID] = bkpLocationName
+				err := CreateCloudCredential(provider, cloudCredName, cloudCredUID, orgID, ctx)
+				log.FailOnError(err, fmt.Sprintf("Verifying creation of cloud credential named [%s] for org [%s] with [%s] as provider", cloudCredName, orgID, provider))
+				err = CreateBackupLocation(provider, bkpLocationName, backupLocationUID, cloudCredName, cloudCredUID, getGlobalBucketName(provider), orgID, "")
+				log.FailOnError(err, fmt.Sprintf("Creating backup location %s", bkpLocationName))
+			}
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/namespace_labelled_backup_test.go:104]: https://portworx.atlassian.net/browse/PA-815

File: tests/backup/namespace_labelled_backup_test.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+			err = CreateSourceAndDestClusters(orgID, "", "", ctx)
+			log.FailOnError(err, "Creating source and destination cluster")
+			srcClusterStatus, err = Inst().Backup.GetClusterStatus(orgID, SourceClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", SourceClusterName))
+			dash.VerifyFatal(srcClusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", SourceClusterName))
+			srcClusterUid, err = Inst().Backup.GetClusterUID(ctx, orgID, SourceClusterName)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster uid", SourceClusterName))
+			destClusterStatus, err = Inst().Backup.GetClusterStatus(orgID, destinationClusterName, ctx)
+			log.FailOnError(err, fmt.Sprintf("Fetching [%s] cluster status", destinationClusterName))
+			dash.VerifyFatal(destClusterStatus, api.ClusterInfo_StatusInfo_Online, fmt.Sprintf("Verifying if [%s] cluster is online", destinationClusterName))
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/namespace_labelled_backup_test.go:119]: https://portworx.atlassian.net/browse/PA-815

File: tests/backup/namespace_labelled_backup_test.go (hunk 1 above)
Comment (snigdha-px) [reply tests/backup/namespace_labelled_backup_test.go:194]: https://portworx.atlassian.net/browse/PA-815

deleting restore namespaces: Tarun is working on it

File: tests/backup/namespace_labelled_backup_test.go (hunk 2 above)
Comment (snigdha-px) [reply tests/backup/namespace_labelled_backup_test.go:1]: done

**Review by skonda-px — CHANGES_REQUESTED** [review]

File: tests/backup/backup_namespace_labelled_test.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+		})
+		// Labelled only half of the namespaces
+		Step("Adding labels to only few namespaces", func() {
+			log.InfoD("Adding labels to only few namespaces")
+			for i := 0; i < numberOfUsers; i++ {
+				err := Inst().S.AddNamespaceLabel(bkpNamespaces[i], labels)
+				log.FailOnError(err, "Failed to add labels %v to namespace %s", labels, bkpNamespaces[i])
+				listOfLabelledNamespaces = append(listOfLabelledNamespaces, bkpNamespaces[i])
+			}
+			log.InfoD("List of label applied namespaces are %v", listOfLabelledNamespaces)
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_namespace_labelled_test.go:83]: done

File: tests/backup/backup_namespace_labelled_test.go [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+			var wg sync.WaitGroup
+			for i := 0; i < numberOfUsers; i++ {
+				backupName := fmt.Sprintf("%s-%v", BackupNamePrefix, users[i])
+				backupNames = append(backupNames, backupName)
+				wg.Add(1)
+				go func(backupName string) {
+					defer GinkgoRecover()
+					defer wg.Done()
+					err := CreateBackupWithNamespaceLabel(backupName, SourceClusterName, bkpLocationName, backupLocationUID,
+						labelSelectors, orgID, srcClusterUid, "", "", "", "", mapToString(labels), listOfLabelledNamespaces, ctx)
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_namespace_labelled_test.go:135]: done

File: tests/backup/backup_namespace_labelled_test.go [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+		log.InfoD("List of all namespaces deployed are %v", bkpNamespaces)
+	})
+	It("Take namespace labelled backup and share with users having different access mode", func() {
+		providers := getProviders()
+		Step("Validate applications", func() {
+			log.Infof("Validate applications")
+			ValidateApplications(contexts)
+		})
+		Step("Generating multiple labels", func() {
+			log.InfoD("Generating multiple labels")
```
[the remark this answers is not in the document]
Comment (snigdha-px) [reply tests/backup/backup_namespace_labelled_test.go:66]: done

**Review by skonda-px — APPROVED** [review]
LGTM

File: tests/backup/backup_helper.go [hunk 9]
```diff
@@ -2026,3 +2029,18 @@ func MapToKeyValueString(m map[string]string) string {
 	}
 	return strings.Join(pairs, ",")
 }
+
+// AreStringSlicesEqual compares two slices of string
+func AreStringSlicesEqual(slice1 []string, slice2 []string) bool {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:2034]: Make this function accept any interface

Comment (snigdha-px) [reply tests/backup/backup_helper.go]: done

File: tests/backup/backup_helper.go [hunk 10]
```diff
@@ -2072,3 +2075,44 @@ func FetchNamespacesFromBackup(ctx context.Context, backupName string, orgID str
 	backedUpNamespaces = resp.GetBackup().GetNamespaces()
 	return backedUpNamespaces, err
 }
+
+// AreSlicesEqual verifies if two slices are equal or not
+func AreSlicesEqual(slice1, slice2 interface{}) bool {
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:2080]: It can be simplified this way - https://go.dev/play/p/jrB5l2T-w-_x

Comment (snigdha-px) [reply tests/backup/backup_helper.go]: done

Comment (snigdha-px) [reply tests/backup/backup_helper.go:2079]: https://jenkins.pwx.dev.purestorage.com/job/Users/job/Sumit/job/Custom-Pipelines/job/px-backup-on-demand-system-test-byoc/1074/consoleFull

**Review by tthurlapati-px — APPROVED** [review]
LGTM

**Comment by mkoppal-px** [comment]
@snigdha-px Why is Environment cleanup failing in your runs?

**Comment by snigdha-px** [comment]
> 

@mkoppal-px 
I have used the same testbed for run which I am using for 2.4.2 testing.
Since my schedule runs/locked bucket are present, it failed to delete

**Review by mkoppal-px — APPROVED** [review]
LGTM

**Pull request merged**